Add `verbose` option

MarkTiedemann opened the pull request:
My first pull request. :)

- Adds a `verbose` option to improve performance (see: [https://github.com/sindresorhus/tasklist/issues/6](https://github.com/sindresorhus/tasklist/issues/6])
- Adds an `apps` option (to return only Windows Store Apps)
- Makes the `filter` option accept not only a filter array, but also a single filter string

MarkTiedemann committed:
Add verbose option, add apps option, and update filter option

MarkTiedemann commented:
CI failed for `nodejs_version=0.12` with the following output:

```
const debug = require('debug')('xo');
^^^^^
SyntaxError: Use of const in strict mode.
```

Should support for Node v0.12 get dropped or should the `xo` devDependency be downgraded?

sindresorhus commented on index.js:
```diff
@@ -18,26 +18,41 @@ module.exports = function (opts) {
+		args.push('/apps');
+	}
+
+	if (Array.isArray(opts.filter)) {
```
I prefer to keep it `Array` only. I don't like mixing types with little benefit.

sindresorhus commented on readme.md:
```diff
@@ -51,6 +46,29 @@ Type: `object`
+
+Type: `boolean`
+
+Return verbose results (default: `false`).
```
Write the default value like done here: https://github.com/sindresorhus/boxen#borderstyle

sindresorhus commented on readme.md:
```diff
@@ -24,12 +24,7 @@ tasklist().then(data => {
-		status: 'Running',
-		username: 'SINDRESORHU3930\\sindre'
-		cpuTime: 0,  // seconds
-		windowTitle: 'Task Host Window'
```
I think it's useful for people to see the properties like this with values, so maybe add another example in the `verbose` option section and show these.

sindresorhus commented:
> Should support for Node v0.12 get dropped or should the xo devDependency be downgraded?

I will do it after this PR. Just ignore Node.js 0.12 failures for now.

MarkTiedemann committed:
Improve pull request

- Make test cases check all properties
- Revert making `memUsage` a `verbose` property (which it is not)
- Revert making `filter` option accept strings
- Add example to `verbose` option documentation
- Change formatting of default values in documentation

MarkTiedemann commented:
I have made the changes you requested and even fixed a bug in my code. :)

- Make test cases check all properties
- Revert making `memUsage` a `verbose` property (which it is not)
- Revert making `filter` option accept strings
- Add example to `verbose` option documentation
- Change formatting of default values in documentation

MarkTiedemann commented:
Just noticed CI failed, and for good reason:
- It seems like `windowTitle` won't be set in the CI env
- The `apps` version returns different rows: 
    - Specifically, it does not have `sessionName` and `sessionNumber`, but instead it has `packageName`
    - In `verbose` mode, it has both `sessionName` and `sessionNumber` as well as `packageName`

Gonna update the PR as soon as possible. :)

MarkTiedemann committed:
Remove apps option

MarkTiedemann committed:
Don't test for windowTitle in CI env

MarkTiedemann commented:
Earlier CI failed on Node 0.12 with linting errors, now it failed on Node 4, too.

sindresorhus commented:
@MarkTiedemann What's the status of this PR? :)

MarkTiedemann commented:
Well, I (almost) forgot about the PR. ;)

But thanks to your question, last night I spent 2-3 hours revisiting, upgrading, re-testing, fixing, and completely refactoring the old code. :)

Not 100% happy with the results yet, and some documentation is still missing, but I'm probably gonna push the results in the next few days.

PS: Last time I worked on this PR a couple of weeks ago, I created a streaming version of this module: [tasklist-stream](https://github.com/MarkTiedemann/tasklist-stream). I think once this PR is done, we could perhaps merge both in some way (they do "share" a lot of code already ...).

MarkTiedemann committed:
Update PR with new language features, bug fixes, and improved tests

This commit includes the following changes:

 - Update Node versions in appveyor.yml
 - Update dependencies
 - Use const instead of var
 - Use () => {} instead of function() {}
 - Remove Promise ponyfill
 - Remove 'N/A' check due to failure on non-English systems
 - Make the tests cases cover all returned tasks

MarkTiedemann commented:
@sindresorhus Here's my refactoring:

 - Update Node versions in appveyor.yml
 - Update dependencies
 - Use const instead of var
 - Use () => {} instead of function() {}
 - Remove Promise ponyfill
 - Remove 'N/A' check due to failure on non-English systems
 - Make the tests cases cover all returned tasks

Hope you like it. :)

sindresorhus commented on test.js:
```diff
@@ -1,22 +1,47 @@
-	t.is(typeof d.pid, 'number');
-	t.is(typeof d.memUsage, 'number');
-});
+const makeTest = (options, t) => {
```
Can you use a [test macro](https://github.com/avajs/ava#test-macros) instead?

MarkTiedemann replied:
Been using ava since .9, I totally missed that you support test macros now. Thanks for the advice!

sindresorhus commented on test.js:
```diff
@@ -1,22 +1,47 @@
+const makeTest = (options, t) => {
+	return tasklist(options).then(tasks => {
+		t.true(tasks.length > 0);
+		tasks.forEach(task => {
```
Use a for-of loop.

sindresorhus commented on test.js:
```diff
@@ -1,22 +1,47 @@
-import test from 'ava';
-import fn from './';
+const test = require('ava');
+const tasklist = require('./');
```
Don't change this.

MarkTiedemann replied:
I didn't like the inconsistency of having require statements in the index file and import statements in the test file. But I guess your plan is to change to import statements everywhere in the long run, so yeah, gladly changing it back. :)

sindresorhus commented on package.json:
```diff
@@ -31,13 +31,13 @@
-    "ava": "*",
-    "xo": "*"
+    "ava": "^0.19.1",
+    "xo": "^0.18.1"
```
Don't change this.

sindresorhus commented on package.json:
```diff
@@ -10,10 +10,10 @@
   },
   "scripts": {
-    "test": "xo && ava"
+    "test": "xo --fix && ava --verbose"
```
Don't change this.

sindresorhus commented on package.json:
```diff
@@ -10,10 +10,10 @@
   },
   "engines": {
-    "node": ">=0.10.0"
+    "node": ">=4.0.0"
```
```
"node": ">=4"
```

sindresorhus commented on test.js:
```diff
@@ -1,22 +1,47 @@
-	t.is(typeof d.memUsage, 'number');
-});
+const makeTest = (options, t) => {
+	return tasklist(options).then(tasks => {
```
Use async/await

sindresorhus commented on index.js:
```diff
@@ -1,56 +1,52 @@
-			args.push('/fi', el);
-		});
+	if (Array.isArray(options.filter)) {
+		options.filter.forEach(fi => args.push('/fi', fi));
```
Use for-of loop.

MarkTiedemann replied:
There was no for-loop before. I'm gladly changing it, but may I ask why?

sindresorhus replied:
It's faster. I'm trying to use it over forEach whenever possible. So for consistency too. And I don't like inline arrow functions for things that doesn't return anything. Makes the code harder to read.

MarkTiedemann replied:
Okay, I understand that. Thanks for explaining!

PS: I have fixed the requested changes. :)

sindresorhus commented on index.js:
```diff
@@ -1,56 +1,52 @@
 	}
 
-	opts = opts || {};
+	const options = opts || {};
```
Don't change this. I have it like this so it will be super easy to change to default parameters when I can target Node.js 6.

MarkTiedemann replied:
I didn't think about that. Thanks for explaining. :)

sindresorhus reviewed (changes requested):
Looks really good. Just some minor nitpick.

MarkTiedemann committed:
Fix requested changes

sindresorhus commented:
Looks great. Thank you @MarkTiedemann :)

sindresorhus commented:
Would you be interested in being a maintainer here? I don't think this module requires much more work, so it's mostly just to highlight your contribution here in the readme.

MarkTiedemann commented:
Yeah, sure. :) Thanks a lot!

Well, there's some work I'd like to do here in the time to come:
- Implement the `/apps` flag
- Implement the `/m` module flag
- Implement the `/svc` service flag
- Make a streaming interface available (I already created a prototype at [tasklist-stream](https://github.com/MarkTiedemann/tasklist-stream), but there's more work to be done there, too. And I'm not sure how to "merge" both modules - if they should be merged at all. 
- As per `tasklist /?`, `windowTitle` and `status` filters are not supported when querying a remote machine. I'd like to implement a check for that, and also test whether this case breaks the code of this module in any way.

sindresorhus commented:
Awesome! Could you put the above into an issue with checklists?

MarkTiedemann commented:
Yeah, sure. See: https://github.com/sindresorhus/tasklist/issues/8.
